File: DemoServer/AdditionalLanguages/python/javascriptIndexes/javascriptMapIndex/javascriptMapIndex.py
from typing import List
from demo_example import Example, RunParamsBase
from models import Employee

#region Usings
from ravendb.documents.indexes.index_creation import AbstractJavaScriptIndexCreationTask
#endregion

#region Demo
#region Step_1
class Employees_ByImportantDetailsJS(AbstractJavaScriptIndexCreationTask):
#endregion

    #region Step_2
    class IndexEntry:
        def __init__(
            self,
            full_name: str = None,
            country: str = None,
            working_in_company_since: int = None,
            number_of_territories: int = None,
        ):
            self.full_name = full_name
            self.country = country
            self.working_in_company_since = working_in_company_since
            self.number_of_territories = number_of_territories
    #endregion

    #region Step_3
    def __init__(self):
        super().__init__()
        self.maps = {
            (
                "map('Employees', function (employee) {"
                "   return {"
                "       full_name: employee.FirstName + ' ' + employee.LastName,"
                "       country: employee.Address.Country,"
                "       working_in_company_since: new Date(employee.HiredAt).getFullYear(),"
                "       number_of_territories: employee.Territories.length"
                "   };"
                "})"
            )
        }
    #endregion
#endregion

class RunParams(RunParamsBase):
    def __init__(self, start_year: int = None):
        self.start_year = start_year

class JavaScriptMapIndex(Example):
    def run(self, run_params: RunParams) -> List[Employee]:
        startYear = run_params.start_year
        Employees_ByImportantDetailsJS().execute(self.document_store_holder.store())

        #region Demo
        with self.document_store_holder.store().open_session() as session:
            #region Step_4
            employees_from_USA = list(
                session.query_index_type(
                    Employees_ByImportantDetailsJS,
                    Employees_ByImportantDetailsJS.IndexEntry,
                )
                .where_equals("country", "USA")
                .where_greater_than("working_in_company_since", startYear)
                .select_fields(Employee)
            )
            #endregion
        #endregion
        
        return employees_from_USA

File: DemoServer/AdditionalLanguages/php/javascriptIndexes/javascriptMapIndex/JavascriptMapIndex.php
<?php

namespace RavenDB\Demo\javascriptIndexes\javascriptMapIndex;

//region using
use RavenDB\Documents\Indexes\AbstractJavaScriptIndexCreationTask;
//endregion

use RavenDB\Demo\common\DocumentStoreHolder;
use RavenDB\Demo\common\models\Employee;

//region Demo
//region Step_1
class Employees_ByImportantDetailsJS extends AbstractJavaScriptIndexCreationTask
//endregion
{
    //region Step_2
    public function __construct()
    {
        parent::__construct();

        $this->setMaps([
                "map('Employees', function (employee) {" .
                "   return {" .
                "       FullName: employee.FirstName + ' ' + employee.LastName," .
                "       Country: employee.Address.Country," .
                "       WorkingInCompanySince: new Date(employee.HiredAt).getFullYear()," .
                "       NumberOfTerritories: employee.Territories.length" .
                "   };" .
                "})"
            ]);
    }
    //endregion
}
//endregion

//region Demo
//region Step_3
class IndexEntry
{
    public ?string $FullName = null;
    public ?string $Country = null;
    public ?int $WorkingInCompanySince = null;
    public ?int $NumberOfTerritories = null;

    public function getFullName(): ?string
    {
        return $this->FullName;
    }

    public function setFullName(?string $FullName): void
    {
        $this->FullName = $FullName;
    }

    public function getCountry(): ?string
    {
        return $this->Country;
    }

    public function setCountry(?string $Country): void
    {
        $this->Country = $Country;
    }

    public function getWorkingInCompanySince(): ?int
    {
        return $this->WorkingInCompanySince;
    }

    public function setWorkingInCompanySince(?int $WorkingInCompanySince): void
    {
        $this->WorkingInCompanySince = $WorkingInCompanySince;
    }

    public function getNumberOfTerritories(): ?int
    {
        return $this->NumberOfTerritories;
    }

    public function setNumberOfTerritories(?int $NumberOfTerritories): void
    {
        $this->NumberOfTerritories = $NumberOfTerritories;
    }
}
//endregion
//endregion

class JavascriptMapIndex
{
    public function __invoke(RunParams $runParams): array
    {
        $startYear = $runParams->getStartYear();

        //region Demo
        $employeesFromUSA = [];
        $session = DocumentStoreHolder::getStore()->openSession();
        
        try {
            //region Step_4
            $employeesFromUSA = $session->query(IndexEntry::class, Employees_ByImportantDetailsJS::class)
                ->whereEquals("Country", "USA")
                ->whereGreaterThan("WorkingInCompanySince", $startYear)
                ->selectFields(Employee::class)
                ->toList();
            //endregion
            
        } finally {
            $session->close();
        }
        //endregion
        
        return $employeesFromUSA;
    }
}
